### PR TITLE
feat(policy): Add domain based and view based policies

### PR DESCRIPTION
**Pull request by dexter-mh-lee**
Add 2 main features

Entity page view based policy
VIEW_ENTITY_PAGE privilege gates access to the entity page as a whole (no gating on graphQL)
VIEW_DATASET_PROFILE privilege gates access to the dataset profile aspect (including graphQL)
VIEW_DATASET_USAGE privilege gates access to the dataset usage statistics endpoint (including graphQL)
Domain based policy
To do so, made some changes to the ResourceFilter
Create a PolicyMatchFilter (GraphQL - Filter in PDL) object that stores a list of criterions to combine. Deprecate the old hardcoded fields like type and resources.
Create a AuthorizationRequestHydrator that defines logic for lazily fetches field values for each field type (ENTITY_TYPE, ENTITY_URN, DOMAIN, etc.) for a given urn. The fetching logic is run only when the field value for the given type is needed to check whether the policy should be applied.

**Note** the new policy format is backward compatible, and on edit, it will convert existing format into the new format. 

Screen Shot 2022-03-28 at 11 58 06 PM

Screen Shot 2022-03-29 at 12 03 33 AM

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/authorization/AuthorizationUtils.java [hunk 1]
```diff
@@ -14,25 +14,25 @@
...
-    return isAuthorized(authorizer, actor, new DisjunctivePrivilegeGroup(ImmutableList.of(andGroup)));
+    return isAuthorized(context, Optional.empty(), PoliciesConfig.GENERATE_PERSONAL_ACCESS_TOKENS_PRIVILEGE);
   }
 
   public static boolean canManageDomains(@Nonnull QueryContext context) {
+    return isAuthorized(context, Optional.empty(), PoliciesConfig.MANAGE_DOMAINS_PRIVILEGE);
+  }
+
+  public static boolean isAuthorized(
+      @Nonnull QueryContext context,
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/authorization/AuthorizationUtils.java:29]: Nice refactor!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/policy/GetGrantedPrivilegesResolver.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.linkedin.metadata.authorization.PoliciesConfig;
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static com.linkedin.datahub.graphql.resolvers.ResolverUtils.bindArgument;
+
+
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/policy/GetGrantedPrivilegesResolver.java:19]: Minor: can we add quick javadoc to this class - i've been trying to add to resolvers

File: datahub-graphql-core/src/main/resources/entity.graphql [hunk 3]
```diff
@@ -5873,6 +5975,51 @@ type ListPoliciesResult {
...
+  """
+  Resource urn
+  """
+  resourceUrn: String!
+}
+
+"""
+Object that encodes the privileges the actor has for a given resource
+"""
+type Privileges {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/entity.graphql:6011]: ResourcePrivileges?

Comment (jjoyce0510) [reply datahub-graphql-core/src/main/resources/entity.graphql:6011]: discussing offline..

File: datahub-web-react/src/app/policy/PolicyPrivilegeForm.tsx [hunk 4]
```diff
@@ -106,82 +122,154 @@ export default function PolicyPrivilegeForm({
...
+            ),
+        });
+    };
+
+    // When a domain is selected, add its urn to the list of domains
+    const onSelectDomain = (domain) => {
+        const filter = resources.filter || {
+            criteria: [],
+        };
+        // urnToDisplayName[domain] = option.children;
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/policy/PolicyPrivilegeForm.tsx:190]: an get rid of this?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/load/TimeSeriesAspectResolver.java [hunk 5]
```diff
@@ -42,6 +48,16 @@ public TimeSeriesAspectResolver(final EntityClient client, final String entityNa
     _aspectMapper = aspectMapper;
   }
 
+  // Check whether the actor is authorized to fetch the timeseries aspect given the resource urn
```
[the remark this answers is not in the document]
Comment (dexter-mh-lee) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/load/TimeSeriesAspectResolver.java:51]: done!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/policy/GetGrantedPrivilegesResolver.java (hunk 2 above)
Comment (dexter-mh-lee) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/policy/GetGrantedPrivilegesResolver.java:19]: done!

File: datahub-web-react/src/app/policy/PolicyPrivilegeForm.tsx (hunk 4 above)
Comment (dexter-mh-lee) [reply datahub-web-react/src/app/policy/PolicyPrivilegeForm.tsx:190]: done!

File: metadata-models/src/main/pegasus/com/linkedin/policy/DataHubResourceFilter.pdl [hunk 6]
```diff
@@ -12,15 +12,20 @@ record DataHubResourceFilter {
...
    * A specific set of resources to apply the policy to, e.g. asset urns
    */
-  resources: optional array[string]
+  resources: optional array[string] // DEPRECATED
 
   /**
    * Whether the policy should be applied to all assets matching the filter.
    */
-   allResources: boolean = false
+  allResources: boolean = false // DEPRECATED
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/policy/DataHubResourceFilter.pdl:25]: We can do in followup - typically we use @deprecated here

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/FieldResolver.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ */
+@RequiredArgsConstructor
+public class FieldResolver {
+  private final Supplier<CompletableFuture<FieldValue>> resolveField;
+  @Getter(lazy = true)
+  private final CompletableFuture<FieldValue> fieldValuesFuture = resolveField.get();
+
+  private static final FieldValue EMPTY = new FieldValue(Collections.emptySet());
+
+  // Helper function that returns FieldResolver for precomputed values
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/FieldResolver.java:26]: Minor change - java docs

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/FilterUtils.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+public class FilterUtils {
+
+  public static final PolicyMatchFilter EMPTY_FILTER =
+      new PolicyMatchFilter().setCriteria(new PolicyMatchCriterionArray());
+
+  private FilterUtils() {
+  }
+
+  // Creates new PolicyMatchCriterion with field and value, using EQUAL PolicyMatchCondition.
```
[the remark this answers is not in the document]
Comment (dexter-mh-lee) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/FilterUtils.java:26]: ah this was copied from the search FilterUtils haha will fix

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldResolverProvider.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import com.datahub.authorization.FieldResolver;
+import com.datahub.authorization.ResourceSpec;
+
+
+/**
+ * Base class for defining a class that provides the field resolver for the given field type
+ */
+public interface ResourceFieldResolverProvider {
+
+  // Field that this hydrator is hydrating
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldResolverProvider.java:12]: Minor: java docs

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldType.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package com.datahub.authorization.fieldresolverprovider;
+
+public enum ResourceFieldType {
```
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldType.java:3]: Minor : java docs on the enum + symbols

File: metadata-service/restli-client/src/main/java/com/linkedin/entity/client/OwnershipClient.java [hunk 11]
```diff
@@ -1,59 +0,0 @@
-package com.linkedin.entity.client;
-
```
Comment (jjoyce0510) [inline metadata-service/restli-client/src/main/java/com/linkedin/entity/client/OwnershipClient.java:2]: HELL YEAH!

File: metadata-service/war/src/main/resources/boot/policies.json [hunk 12]
```diff
@@ -38,143 +38,30 @@
         ]
       },
       "privileges":[
-        "EDIT_ENTITY"
+        "EDIT_ENTITY",
```
Comment (jjoyce0510) [inline metadata-service/war/src/main/resources/boot/policies.json:41]: This one will be fun to merge :p

Comment (dexter-mh-lee) [reply metadata-service/war/src/main/resources/boot/policies.json:41]: dreading it

File: metadata-models/src/main/pegasus/com/linkedin/policy/DataHubResourceFilter.pdl (hunk 6 above)
Comment (dexter-mh-lee) [reply metadata-models/src/main/pegasus/com/linkedin/policy/DataHubResourceFilter.pdl:25]: done

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/FieldResolver.java (hunk 7 above)
Comment (dexter-mh-lee) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/FieldResolver.java:26]: done

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldResolverProvider.java (hunk 9 above)
Comment (dexter-mh-lee) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldResolverProvider.java:12]: done

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldType.java (hunk 10 above)
Comment (dexter-mh-lee) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/ResourceFieldType.java:3]: done

File: datahub-web-react/src/app/policy/PolicyDetailsModal.tsx [hunk 13]
```diff
@@ -98,26 +110,50 @@ export default function PolicyDetailsModal({ policy, visible, onClose, privilege
...
-                                );
-                            })}
-                            {policy?.resources?.allResources && <PoliciesTag>All</PoliciesTag>}
+                            {(resourceEntities?.length &&
+                                resourceEntities.map((value, key) => {
+                                    return (
+                                        // eslint-disable-next-line react/no-array-index-key
+                                        <PoliciesTag key={`resource-${value.value}-${key}`}>
+                                            <Typography.Text>
+                                                {getDisplayName(value.entity) || value.value}
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/policy/PolicyDetailsModal.tsx:137]: minor : 

do value.entity && getDisplayName(value.entity) || value.value (in both)

**Review by jjoyce0510 — APPROVED** [review]
Let's make sure this passes CI - then we are good to go.

**Pull request merged**